# PopDel-scripts
Collection of scripts used for the paper
Sebastian Niehus, Hákon Jónsson, Janina Schönberger, Eythór Björnsson, Doruk Beyter, Hannes P. Eggertsson, Patrick Sulem, Kári Stefánsson, Bjarni V. Halldórsson, Birte Kehr.
*PopDel identifies medium-size deletions jointly in tens of thousands of genomes*.  
Available at Nature Communications: https://www.nature.com/articles/s41467-020-20850-5  
All generated output VCF/BCF files are published at Zenodo; DOI: https://doi.org/10.5281/zenodo.3992607

**Content**

This repository contains the scripts used for the comparisons and evaluations of the paper *PopDel calls deletions jointly in tens of thousands of genomes*. It does not contain the public data of of the [Polaris HiSeqX Diversity Cohort](https://github.com/Illumina/Polaris/wiki/HiSeqX-Diversity-Cohort), [Polaris Kids Cohort](https://github.com/Illumina/Polaris/wiki/HiSeqX-Kids-Cohort) or the [Illumina Platinum Genome NA12878](https://www.ebi.ac.uk/ena/data/view/ERR194147). They can be obtained from the respective online sources. Running the scripts requires you to adapt the paths in some of them.

 - [Random Deletion Data Simulation](#random-deletion-data-simulation)
 - [G1k Deletion Data Simulation](#g1k-deletion-data-simulation)
 - [Call Set comparison of Different Variant Callers on HG002 Trio](#call-set-comparison-of-different-variant-callers-on-hg002-trio)
 - [Call Set comparison of Different Variant Callers on NA12878](#call-set-comparison-of-different-variant-callers-on-na12878)
 - [Polaris HiSeqX Diversity Cohort](#polaris-hiseqx-diversity-cohort)
 - [Polaris Kids Cohort](#polaris-kids-cohort)

## Random Deletion Data Simulation

 [*Simulation/uniform_simulation/deletion_data_Simulation/uniform_simulation/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/deletion_data_simulation)
- [simulate_deletion_haplotype.py](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/deletion_data_simulation/simulate_deletion_haplotype.py): Simulation of haplotypes.
- [simulate_deletions.py](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/deletion_data_simulation/simulate_deletions.py): Simulation of reads of the samples
- [simulate.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/deletion_data_simulation/simulate.sh): Wrapping above scripts and aligning the simulated reads to the GRCh38.
- [simulation2vcf.py](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/deletion_data_simulation/simulation2vcf.py): Writing the simulated deletions to VCF-files.

Note on random seeds: We used 0 as random seed to simulate 2000 deletions with the script 'simulate_deletions.txt'. The random seed used for the simulation of reads using 'simulate.sh' was 1 for the first sample, 2 for the the second sample and so on.

 [*Simulation/uniform_simulation/delly/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/delly)
 - [generateDelly.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/delly/generateDelly.sh): generates the scripts for each batch size.
 - [runDelly.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/delly/runDelly.sh): Runs the scripts generated by generateDelly.sh and measures the resource consumption.

 [*Simulation/uniform_simulation/gridss/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/gridss)
 - [generateGridss.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/gridss/generateGridss.sh): Generates the scripts for each batch size.
 - [runGridss.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/gridss/runGridss.sh): Calls all the scripts generated by generateGridss.sh and measures the resource consumption.
 - [gridss.filter.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/gridss/gridss.filter.sh): Deduplicates the break ends called by GRIDSS and annotates the VCFs.
 - [annotate.R](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/gridss/annotate.R): Annotation script called by gridss.filter.sh.

 [*Simulation/uniform_simulation/smoove/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/smoove)
 - [run_smoove.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/run_smoove.sh): Runs the scripts [smoove_single.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/smoove_single.sh), [smoove_1.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/smoove_1.sh), [smoove_2.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/smoove_2.sh), [smoove_3.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/smoove_3), [smoove_4.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/smoove_4.sh) and measures the resource consumption.
 - [smoove_filter.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/smoove_filter.sh): Filters Smoove's call set.
 - [smoove.env](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/smoove.env): Contains the conda environment for Smoove.
 - [exclude.bed](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/smoove/exclude.bed): Contains the regions Smoove should ignore.

 [*Simulation/uniform_simulation/podel/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/popdel)
- [popdelProfile.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/popdel/popdelProfile.sh): Contains the commands for creating a profile of each bam file.
- [runPopDel.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/popdel/runPopDel.sh): Runs all the scripts and measures the resource consumption.

[*Simulation/uniform_simulation/truth/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/truth)
Contains an archive ([truth.tar.gz](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/truth/truth.tar.gz)) of all simulated variants for each batch size used for the evaluation. The files are the results of the deletion data simulation with above mentioned random seeds.

[*Simulation/uniform_simulation/plots/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/plots)
- [eval_bed.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/plots/eval_bed.sh): Evaluates the TP/FP/FN of all tools. Note that the range of the evaluation loop might have to be adjusted to match the batch sizes that the respective tools actually processed successfully.
- [compare_results.py](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/plots/compare_results.py): Script for alternative evaluation, based on fixed positional and size-estimate margins. can also consider genotypes of individual samples.
- [eval.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/plots/eval.sh): Wrapper for [compare_results.py](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/plots/compare_results.py)
- [simulation_plots.R](https://github.com/kehrlab/PopDel-scripts/blob/master/Simulation/uniform_simulation/plots/simulation_plots.R): Script for generating all plots of the simulated data.

## G1k Deletion Data Simulation

[*Simulation/uniform_simulation/deletion_data_Simulation/g1k_simulation/*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation)
- [*Simulation/uniform_simulation/deletion_data_Simulation/g1k_simulation/additionalFiles*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/additionalFiles): Contains the G1k deletion reference set and exclusion files for smoove.
- [*Simulation/uniform_simulation/deletion_data_Simulation/g1k_simulation/environments*](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/environments): Contains the conda environments for Delly, Manta and Smoove.
- [Snakefile.wholeGenome.deletion.simulation](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/Snakefile.wholeGenome.deletion.simulation): Snakefile for use with Snakemake. Manages the complete simulation and calling workflow for all tools.
- [config.yml](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/config.yml): Configuration for the corresponding Snakefile.
- [eval_full.sh](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/eval_full.sh): Compares the VCFs of predicted deletions with the simulated deletions.
- [simulation_g1k_plots.R](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/simulation_g1k_plots.R): Takes the tables created by [eval_full.sh](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/eval_full.sh) and plots the results and creates additional tables for comparisons.
- [simulate_deletion_haplotype.py](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/simulate_deletions.py): Simulates a variant haplotype sequence from a given deletion. Used by [Snakefile.wholeGenome.deletion.simulation](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/Snakefile.wholeGenome.deletion.simulation).
- [simulation2vcf.py](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/simulation2vcf.py): Writes the simulated variants in VCF format. Used by [Snakefile.wholeGenome.deletion.simulation](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/Snakefile.wholeGenome.deletion.simulation).
- [compare_results.py](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/compare_results.py): Compares the predicted to simulated variants for on sample. Used by [eval_full.sh](https://github.com/kehrlab/PopDel-scripts/tree/master/Simulation/uniform_simulation/g1k_simulation/eval_full.sh).

## Call Set comparison of Different Variant Callers on HG002 Trio

[*Ashkenazim_trio*](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/delly)
- [delly.environment.yml](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/delly/delly.environment.yml): Contains the conda environment for Delly.
- [Snakefile_call.delly.giab](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/delly/Snakefile_call.delly.giab): Snakefile to be used with Snakemake to manage Delly's workflow.
- [config.yml](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/delly/config.yml): Configuration for the corresponding Snakefile.
- [filter_delly.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/delly/filter_delly.sh): Filters Delly's call sets.
- [human.hg19.excl.tsv](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/delly/human.hg19.excl.tsv): Contains the regions Delly should ignore.

[*Ashkenazim_trio*](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/manta)
- [manta.env](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/manta/manta.env): Contains the conda environment for Manta.
- [Snakefile_manta.giab](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/manta/Snakefile_manta.giab): Snakefile to be used with Snakemake to manage Manta's workflow.
- [config.yml](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/manta/config.yml): Configuration for the corresponding Snakefile.
- [filter_manta.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/manta/filter_manta.sh): Filters Manta's call sets.

[*Ashkenazim_trio*](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/popdel)
- [Snakefile_GIAB_popdel](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/popdel/Snakefile_GIAB_popdel): Snakefile to be used with Snakemake to manage Manta's workflow.
- [config.yml](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/popdel/config.yml): Configuration for the corresponding Snakefile.
- [filter_manta.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/popdel/filter_popdel.sh): Filters PopDel's call sets.
- [sampling.regions](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/popdel/sampling.regions): Contains the regions PopDel uses for sampling the background distribution. If the option "-r grch37" is used, this file is not required and the sampling is performed on the same regions.
- [maxCov.tsv](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/popdel/maxCov.tsv): File containing the desired maximum coverage for PopDel to consider for each of the three genomes. The selected values correspond to 3x each samples mean coverage.
- [all.GRCh37.profiles](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/popdel/all.GRCh37.profiles): File containing the paths to the profiles of the three genomes.

[*Ashkenazim_trio*](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/plots/)
- [mendelianError](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/plots/mendelianError): Contains scripts for calculation and plotting of the Mendelian inheritance error.
- [precision_recall](https://github.com/kehrlab/PopDel-scripts/tree/master/Ashkenazim_trio/plots/precision_recall): Contains the scripts for calculation and plotting of precision-recall curve
- [venn_diagram](https://github.com/kehrlab/PopDel-scripts/blob/master/Ashkenazim_trio/plots/venn_diagram):  Contains scripts for calculation and plotting of the venn diagrams.

## Call Set comparison of Different Variant Callers on NA12878

[*NA12878/delly/*](https://github.com/kehrlab/PopDel-scripts/tree/master/NA12878/delly)
- [delly.environment.yml](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/delly/delly.environment.yml): Containing the conda environment for Delly.
- [Snakefile_call.delly.platinum](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/delly/Snakefile_call.delly.platinum): Snakefile to be used with Snakemake to manage Delly's workflow.
- [human.hg38.excl.tsv](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/delly/human.hg389.excl.tsv): Contains the regions Delly should ignore.

- [manta.environment.yml](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/manta/manta.environment.yml): Containing the conda environment for Manta.
- [Snakefile_manta.platinum](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/manta/Snakefile_manta.platinum): Snakefile to be used with Snakemake to manage Manta's workflow.
- [config.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/manta/config.yaml): Configuration for the corresponding Snakefile.
- [GRCh38.regions.bed.gz](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/manta/GRCh38.regions.bed.gz): Contains the regions Manta should ignore.
- [GRCh38.regions.bed.gz.tbi](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/manta/GRCh38.regions.bed.gz.tbi): Index of [GRCh38.regions.bed.gz](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/manta/GRCh38.regions.bed.gz).

[*NA12878/smoove/*](https://github.com/kehrlab/PopDel-scripts/tree/master/NA12878/smoove)
- [Snakefile_call.smoove.platinum](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/smoove/Snakefile_call.smoove.platinum): Snakefile to be used with Snakemake to manage Smoove's workflow.
- [config.yml](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/smoove/config.yml): Configuration for the corresponding Snakefile.
- [smoove.env](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/smoove/smoove.env): File containing the conda environment for Smoove.
- [exclude.cnvnator_100bp.GRCh38.20170403.bed](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/smoove/exclude.cnvnator_100bp.GRCh38.20170403.bed): Contains the regions Smoove should ignore.

[*NA12878/popdel/*](https://github.com/kehrlab/PopDel-scripts/tree/master/NA12878/popdel)
- [profile.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/popdel/profile.sh): Contains the command used for generating the profile of NA12878.
- [platinum.profiles](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/popdel/platinum.profiles): Contains the location of the profile generated by above command. Used as input for PopDel call.
- [Snakefile_call.popdel.platinum](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/popdel/Snakefile_call.popdel.platinum): Snakefile to be used with Snakemake to manage the PopDel call commands for each chromosome.

[*NA12878/reference*](https://github.com/kehrlab/PopDel-scripts/tree/master/NA12878/reference)
- [contromeres.bed](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/reference/centromeres.bed)
- [pacbio/remapped_NA12878_pacbio_deduplicated_deletions.sort.all.bed](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/reference/pacbio/remapped_NA12878_pacbio_deduplicated_deletions.sort.all.bed): Set of high quality reference deletions for NA12878 based on PacBio long reads. Based on variants from [GiaB](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/NA12878/NA12878_PacBio_MtSinai/).
- [personalis/remapped_NA12878_personalis_deduplicated_deletions.sort.all.bed](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/reference/personalis/remapped_NA12878_personalis_deduplicated_deletions.sort.all.bed): Set of high quality reference deletions for NA12878 based on Illumina short reads. Based on variants from [GiaB](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/NA12878/analysis/Illumina_PlatinumGenomes_NA12877_NA12878_09162015/hg38/2.0.1/NA12878/).

[*NA12878/plots*](https://github.com/kehrlab/PopDel-scripts/tree/master/NA12878/plots)
- [VennDiagNA12878.R](https://github.com/kehrlab/PopDel-scripts/tree/master/NA12878/plots/VennDiagNA12878.R): Creates Venn diagrams for the call sets of PopDel, Delly and Lumpy for NA12878. Works on the output of [bedtools/Snakefile.intersect](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/plots/bedtools/Snakefile.intersect)

[*NA12878/plots/bedtools/*](https://github.com/kehrlab/PopDel-scripts/tree/master/NA12878/plots/bedtools)
- [Snakefile.intersect](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/plots/bedtools/Snakefile.intersect): Snakefile for use with Snakemake. Manages the BED-conversion and overlap calculations via bedtools intersect.
- [config.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/NA12878/plots/bedtools/config.yaml): Configuration of the evaluation.

## Polaris HiSeqX Diversity Cohort

[*polaris_diversity_cohort/delly/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_diversity_cohort/delly)
- [config.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/delly/config.yaml): Configuration for the corresponding Snakefile.
- [environment.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/delly/environment.yaml): Contains the conda environment for Delly.
- [Snakefile_polaris_delly](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/delly/Snakefile_polaris_delly): Snakefile for use with Snakemake. Manages Delly's workflow.
- [human.hg38.excl.tsv](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/delly/human.hg38.excl.tsv): Contains the regions Delly should ignore.

[*polaris_diversity_cohort/smoove/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_diversity_cohort/smoove)
- [config.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/smoove/config.yaml): Configuration for corresponding Snakefile.
- [smoove.env](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/smoove/smoove.env): Contains the conda environment for Smoove.
- [Snakefile_smoove_polaris150](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/smoove/Snakefile_smoove_polaris150): Snakefile for use with Snakemake. Manages Smoove's workflow.
- [exclude.cnvnator_100bp.GRCh38.20170403.bed](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/smoove/exclude.cnvnator_100bp.GRCh38.20170403.bed): Contains the regions Smoove should ignore..

[*polaris_diversity_cohort/popdelProfile/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_diversity_cohort/popdelProfile)
- [config.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/popdelProfile/config.yaml): Configuration for corresponding Snakefile.
- [Snakefile_polaris_profile](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/popdelProfile/Snakefile_polaris_profile): Snakefile for use with Snakemake. Manages the creation of the profiles for all samples.

[*polaris_diversity_cohort/popdelCall/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_diversity_cohort/popdelCall)
- [Snakefile_rnd_150](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/popdelCall/Snakefile_rnd_150): Snakefile for use with Snakemake. Applies PopDel call on each chromosome of all samples jointly.

[*polaris_diversity_cohort/plots/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_diversity_cohort/plots)
- [extract_calls.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/plots/extract_calls.sh): Commands for transforming the calls of the tools to the allele-count matrix required by pca_boxplot_varCount.R for the PCA.
- [extract_popdel_per_sample_variants_GT26.sh](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_diversity_cohort/plots/extract_popdel_per_sample_variants_GT26.sh): Counts PopDel's deletions per sample for the given genotype quality (26).
- [ancestry.csv](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_diversity_cohort/plots/ancestry.csv): Lists the ancestry of each sample.
- [pca_boxplot_varCount.R](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_diversity_cohort/plots/pca_boxplot_varCount.R): Script for creating the PCA-plots for all tools and the box-plot and variant counts for PopDel

## Polaris Kids Cohort

[*polaris_kids_cohort/delly/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_kids_cohort/delly)
- [environment.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/delly/environment.yaml): Conda environment for Delly.
- [config.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/delly/config.yaml): Contains the configuration for the corresponding Snakefile.
- [Snakefile_polaris_kids_delly](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/delly/Snakefile_polaris_kids_delly): Manages Delly's workflow.

[*polaris_kids_cohort/smoove/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_kids_cohort/smoove)
- [smoove.env](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/smoove/environment.yml): Conda environment for Smoove.
- [config.yaml](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/smoove/config.yaml): Contains the configuration for the corresponding Snakefile.
- [Snakefile_smoove_polarisKids](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/smoove/Snakefile_polaris_kids_lumpy): Manages Smoove's workflow.
- [exclude.cnvnator_100bp.GRCh38.20170403.bed](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/smoove/exclude.cnvnator_100bp.GRCh38.20170403.bed): Contains the regions Smoove should ignore.

[*polaris_kids_cohort/popdel/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_kids_cohort/popdel)
- [family_kids.profiles](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/popdel/family_kids.profiles): Shuffled list of paths to profiles created by PopDel profile.
- [Snakefile_polaris_kids_popdel](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/popdel/Snakefile_polaris_kids_popdel)Snakefile_polaris_kids: Snakefile for use with Snakemake. Manages PopDel's workflow.

[*polaris_kids_cohort/plots/*](https://github.com/kehrlab/PopDel-scripts/tree/master/polaris_kids_cohort/plots)
- [hwe.py](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/plots/hwe.py): Filters the VCF files according to the Hardy-Weinberg-Equilibrium.
- [transmission_ntrio.py](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/plots/transmission_ntrio.py): Calculates the transmission rates and Mendenlian inheritance error rates on the HWE-filtered VCFs.
- [kids.ped](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/plots/kids.ped): Contains the pedigree information of the kids cohort. One trio <Parent1, Parent2, Child> per line.
- [tr-mendel.R](https://github.com/kehrlab/PopDel-scripts/blob/master/polaris_kids_cohort/plots/tr-mendel.R): Script for creating the transmission rate plots and plots of Mendelian inheritance error.
